### ISSUE TITLE
Coder does not use the installed Terraform version

### ISSUE THREAD

**Issue by 0xmeyer** [coder/coder#12392]
Coder: 2.8.2
Terraform: 1.6.6

Coder runs in a self-built Container, which is based on an Python Image, in which Terraform is also installed. In my previous deployment (Coder 2.6.0) it still worked as usual.

```bash
$ coder --version
Coder v2.8.2+c4a01a4 Mon Feb 12 05:33:05 UTC 2024
https://github.com/coder/coder/commit/c4a01a42ceeb8857856c2204e4d4907f0f93ae5f

Full build of Coder, supports the server subcommand.

$ terraform version -json
{
  "terraform_version": "1.6.6",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": true
}

$ find / -type f -executable -name terraform -print -exec {} version -json \; 2>/dev/null
/home/coder/.cache/coder/provisioner-0/tf/terraform
{
  "terraform_version": "1.4.6",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": true
}
/home/coder/.cache/coder/provisioner-2/tf/terraform
{
  "terraform_version": "1.4.6",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": true
}
/home/coder/.cache/coder/provisioner-1/tf/terraform
{
  "terraform_version": "1.4.6",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": true
}
/usr/bin/terraform
{
  "terraform_version": "1.6.6",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": true
}
```

![Screenshot from 2024-02-27 14-39-23](https://github.com/coder/coder/assets/125983009/baa56ed4-d891-440d-9d29-76836142dde4)

Maybe I'm missing something, but it worked before.

**Comment by ericpaulsen** [comment]
hi @0xmeyer - this is expected. Coder will not use your local Terraform binary, but instead use the binary it installs and caches into `/home/coder/.cache/coder/provisioner-2/tf/terraform`.

**Comment by 0xmeyer** [comment]
Hmm ... I thought a local Terraform binary is preferred over the one installed by Coder, which is why Terraform version 1.6.6 is installed in the [Dockerfile](https://github.com/coder/coder/blob/main/scripts/Dockerfile.base). Or rather I thought that Coder does not install a Terraform binary if one already exists.

**Comment by ericpaulsen** [comment]
Coder will default to the cached Terraform binary in `/home/coder/.cache/coder/provisioner/tf/terraform` if present - hence why it is using `1.4.6`. if you delete this cache, and re-deploy Coder, it will pull down the `1.6.6` binary set in the Dockerfile.

**Comment by 0xmeyer** [comment]
Coder runs in a self-built container, which is then deployed via the Coder Helm Chart in Kubernetes. There is no pv mounted to $HOME. The container image itself does not have a cache directory, which means that when I restart the Kubernetes pod, the cache directory does not exist.

After restarting the deployment, the cache directory is created and Coder seems to pull the Terraform binary (v1.4.6).

```bash
$ podman run -it --entrypoint bash coder "stat /home/coder/.cache"
bash: stat /home/coder/.cache: No such file or directory

$ kubectl exec -it coder-5685865858-zvrmp -- stat /home/coder/.cache
  File: /home/coder/.cache
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: 0,272   Inode: 1700684     Links: 3
Access: (0700/drwx------)  Uid: ( 1000/   coder)   Gid: ( 1000/   coder)
Access: 2024-03-05 06:51:23.845234487 +0100
Modify: 2024-03-05 06:51:09.828998089 +0100
Change: 2024-03-05 06:51:09.828998089 +0100
 Birth: 2024-03-05 06:51:09.828998089 +0100

kubectl rollout restart deployment coder

kubectl exec -it coder-65c8b775bf-hhpnk -- stat /home/coder/.cache
  File: /home/coder/.cache
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: 0,292   Inode: 1700744     Links: 3
Access: (0700/drwx------)  Uid: ( 1000/   coder)   Gid: ( 1000/   coder)
Access: 2024-03-05 07:00:28.778423895 +0100
Modify: 2024-03-05 07:00:19.586268908 +0100
Change: 2024-03-05 07:00:19.586268908 +0100
 Birth: 2024-03-05 07:00:19.586268908 +0100

$ kubectl exec -it coder-65c8b775bf-hhpnk -- /home/coder/.cache/coder/provisioner-0/tf/terraform -version
Terraform v1.4.6
on linux_amd64

Your version of Terraform is out of date! The latest version
is 1.7.4. You can update by downloading from https://www.terraform.io/downloads.html
```

**Comment by ericpaulsen** [comment]
@0xmeyer - then it appears there is an issue with our Dockerfile. cc: @deansheather

**Comment by deansheather** [comment]
Women you say self built container, you mean you use a custom image with our official binary? What's your build process?

Is terraform in the PATH in the container (check using exec)?

**Comment by 0xmeyer** [comment]
> Women you say self built container, you mean you use a custom image with our official binary? 

 ... ????

Sorry for the wrong wording or inaccurate problem description....

Yes, I'm using a **custom** image with your official binary, which I be build using `kaniko`.

```Dockerfile
FROM python3

ENV CODER_VERSION=2.8.2
ENV TERRAFORM_VERSION=1.6.6

ARG DEBIAN_FRONTEND=noninteractive

RUN groupadd -g 1000 coder \
    && useradd --create-home --shell /bin/bash --gid 1000 coder

WORKDIR /tmp

RUN set -x \
    && apt-get update \
    && apt-get install --assume-yes ldap-utils jq git unzip openssl openssh-client python3-ldap3 \
    && rm -rf /var/cache/apt/archives/*deb /var/cache/apt/*pkgcache.bin /var/lib/apt/lists/*

# install coder from GitHub <https://github.com/coder/coder/releases>
RUN apt-get update \
    && curl -qOLs https://github.com/coder/coder/releases/download/v${CODER_VERSION}/coder_${CODER_VERSION}_linux_amd64.tar.gz \
    && curl -qOLs https://github.com/coder/coder/releases/download/v${CODER_VERSION}/coder_${CODER_VERSION}_checksums.txt \
    && grep coder_${CODER_VERSION}_linux_amd64.tar.gz coder_${CODER_VERSION}_checksums.txt | sha256sum -c - \
    && tar -xvf coder_${CODER_VERSION}_linux_amd64.tar.gz \
    && mv coder /usr/bin/ \
    && rm -rf /var/cache/apt/archives/*deb /var/cache/apt/*pkgcache.bin /var/lib/apt/lists/* /tmp/coder* /tmp/{LICENSE,README}*

# install Terraform
RUN apt-get update \
    && curl -qOLs https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_linux_amd64.zip \
    && curl -qOLs https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_SHA256SUMS \
    && grep linux_amd64 terraform_${TERRAFORM_VERSION}_SHA256SUMS | sha256sum -c - \
    && unzip terraform_${TERRAFORM_VERSION}_linux_amd64.zip \
    && chmod 0755 terraform \
    && mv terraform /usr/local/bin/ \
    && rm -rf /var/cache/apt/archives/*deb /var/cache/apt/*pkgcache.bin /var/lib/apt/lists/* /tmp/terraform*

COPY coder-ldapsearch /usr/local/bin/

ENV HOME=/home/coder
USER 1000:1000

WORKDIR /home/coder

ENTRYPOINT [ "/usr/bin/coder", "server" ]
```

> Is terraform in the PATH in the container (check using exec)?

```bash
coder@9426c8d611ea:~$ echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

coder@9426c8d611ea:~$ exec terraform version
Terraform v1.6.6
on linux_amd64

Your version of Terraform is out of date! The latest version
is 1.7.4. You can update by downloading from https://www.terraform.io/downloads.html
```

**Comment by deansheather** [comment]
Sorry, autocorrect. I meant to say "when you say" not "women you say" 🤦

**Comment by 0xmeyer** [comment]
Fixed with Version `2.9.1` 

```bash
$ coder --version
Coder v2.9.1+2a2fd70 Tue Mar 19 12:37:21 UTC 2024
https://github.com/coder/coder/commit/2a2fd706b5b45e48788bf269a93d4c592ed0eec3

Full build of Coder, supports the server subcommand.

$ find / -type f -executable -name terraform -print -exec {} version -json \; 2>/dev/null
/usr/bin/terraform
{
  "terraform_version": "1.6.6",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": true
}
```
